src/realm/views.cpp

```diff
@@ -190,9 +198,121 @@ bool SortDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
     return total_ordering ? i.index_in_view < j.index_in_view : 0;
 }
 
-void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& distinct)
+void DescriptorOrdering::emplace_sort(SortDescriptor&& sort)
 {
-    if (!order && !distinct)
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        m_descriptors.emplace_back(std::move(sort));
+    }
+    else if (descriptor_is_sort(orig_length - 1)) {
+        // merge them together by appending internally
+        m_descriptors[orig_length - 1].merge_with(std::move(sort));
+    }
+    else { // last descriptor is distinct, just add it
+        m_descriptors.emplace_back(std::move(sort));
+    }
+}
+
+void DescriptorOrdering::emplace_distinct(DistinctDescriptor&& distinct)
```

Comment:
It's a bit unusual only to provide && versions. If you instead take `distinct` by value, it will be moved if you pass an rvalue, so you get nearly the same performance. And at the same time allow consts and lvalues to be passed